From 461891c11ca49f4bb0ba68d1d7bf702e221e9629 Mon Sep 17 00:00:00 2001
From: vipinc <vipinc@ami.com>
Date: Tue, 18 Jun 2024 13:48:33 +0530
Subject: [PATCH] Add Configurable FRU ID Support

This commit add support to expose FRU configuration file to D-Bus

Tested in intel Archercity:
- Verified the eeprom json configuration is exposed into D-Bus
Json File:
root@AMIOT-6ad6c0942882:~# cat /usr/share/entity-manager/configurations/eeprom.json
    {
        "FRU_EEPROM": [
            {
                "Bus": 7,
                "Address": 80,
                "FruId": 7,
                "Name": "PSU1_FRU"
            },
            {
                "Bus": 15,
                "Address": 80,
                "FruId": 15,
                "Name": "AVC_LRP"
            },
            {
                "Bus": 0,
                "Address": 0,
                "FruId": 15,
                "Name": "AVC_NLRP"
            }
        ]
  }

D-Bus Data:
    root@AMIOT-6ad6c0942882:~# busctl introspect xyz.openbmc_project.FruDevice /xyz/openbmc_project/FruDevice/PSU1_FRU
    NAME                                         TYPE      SIGNATURE RESULT/VALUE FLAGS
    org.freedesktop.DBus.Introspectable          interface -         -            -
    .Introspect                                  method    -         s            -
    org.freedesktop.DBus.Peer                    interface -         -            -
    .GetMachineId                                method    -         s            -
    .Ping                                        method    -         -            -
    org.freedesktop.DBus.Properties              interface -         -            -
    .Get                                         method    ss        v            -
    .GetAll                                      method    s         a{sv}        -
    .Set                                         method    ssv       -            -
    .PropertiesChanged                           signal    sa{sv}as  -            -
    xyz.openbmc_project.Inventory.Item.FruConfig interface -         -            -
    .Address                                     property  y         80           emits-change
    .Bus                                         property  y         7            emits-change
    .FruId                                       property  y         7            emits-change
    root@AMIOT-6ad6c0942882:~#
    CTRL-A Z for help | 115200 8N1 | NOR | Minicom 2.7.1 |

 Fix-For-Fru-Empty-Configuration

Description:
It will only scan i2c devices that are not mounted as EEPROM files. However, for devices that have already been mounted, it was not checked whether they are in the set FRU list.

Processing functions for mounted EEPROMs in findI2CEproms

Tested:

Build Successfully and Verified,its Working As Expected.

Before Fix:
After Clear the FRU:

root@AMIOT-5242e7a2bc72:~# ipmitool fru list
FRU Device Description : Builtin FRU Device (ID 0)
 Chassis Type          : Rack Mount Chassis
 Chassis Part Number   : ..................
 Chassis Serial        : ..................
 Chassis Extra         : ...............................
 Chassis Extra         : ...............................
 Chassis Area Checksum : OK
 Board Mfg Date        : Mon Jan  1 00:00:00 1996
 Board Mfg             : Intel Corporation
 Board Product         : ArcherCity
 Board Serial          : Intel_3245
 Board Part Number     : ..........
 Board Area Checksum   : OK
 Product Manufacturer  : Intel Corporation
 Product Name          : ArcherCity
 Product Part Number   : ..........
 Product Version       : ....................
 Product Serial        : Intel_3245
 Product Asset Tag     : ....................
 Product Area Checksum : OK

After Fix:

After Clear the Fix:

root@AMIOT-5242e7a2bc72:~# ipmitool fru list
FRU Device Description : Builtin FRU Device (ID 0)
 Chassis Type          : Rack Mount Chassis
 Chassis Part Number   : ..................
 Chassis Serial        : ..................
 Chassis Extra         : ...............................
 Chassis Extra         : ...............................
 Chassis Area Checksum : OK
 Board Mfg Date        : Mon Jan  1 00:00:00 1996
 Board Mfg             : Intel Corporation
 Board Product         : ArcherCity
 Board Serial          : Intel_3245
 Board Part Number     : ..........
 Board Area Checksum   : OK
 Product Manufacturer  : Intel Corporation
 Product Name          : ArcherCity
 Product Part Number   : ..........
 Product Version       : ....................
 Product Serial        : Intel_3245
 Product Asset Tag     : ....................
 Product Area Checksum : OK

FRU Device Description : UNKNOWN (ID 7)
 Unknown FRU header version 0xff

Working as excepted.

Signed-off-by: vipinc <vipinc@ami.com>
---
 src/fru_device.cpp | 52 ++++++++++++++++++++++++++++++++++++++++++++--
 1 file changed, 50 insertions(+), 2 deletions(-)

diff --git a/src/fru_device.cpp b/src/fru_device.cpp
index b4e13b6..dd1774a 100644
--- a/src/fru_device.cpp
+++ b/src/fru_device.cpp
@@ -478,7 +478,8 @@ std::set<size_t> findI2CEeproms(int i2cBus,
         foundList.insert(address);
 
         std::vector<uint8_t> device = processEeprom(i2cBus, address);
-        if (!device.empty())
+        if (!device.empty() ||
+            (device.empty() && isFruConfigured(i2cBus, address)))
         {
             devices->emplace(address, device);
         }
@@ -1069,6 +1070,48 @@ bool writeFRU(uint8_t bus, uint8_t address, const std::vector<uint8_t>& fru)
     close(file);
     return true;
 }
+void DoFruConfig(sdbusplus::asio::object_server& objServer)
+{
+    std::string configFile =
+        "/usr/share/entity-manager/configurations/eeprom.json";
+    using Json = nlohmann::json;
+    std::ifstream jsonFile(configFile);
+    if (!jsonFile.good())
+    {
+        std::cerr << "JSON file not found " << configFile << "\n";
+    }
+
+    Json data = nullptr;
+    try
+    {
+        data = Json::parse(jsonFile, nullptr, false);
+    }
+    catch (const Json::parse_error& e)
+    {
+        std::cerr << "Faile to parse FRU Json file";
+    }
+
+    if (data.empty())
+    {
+        std::cerr << "FRU Config data is empty";
+    }
+    for (const auto& fruConfigs : data["FRU_EEPROM"])
+    {
+        uint8_t busConf = static_cast<uint8_t>(fruConfigs["Bus"]);
+        uint8_t addrConf = static_cast<uint8_t>(fruConfigs["Address"]);
+        std::string fruName = static_cast<std::string>(fruConfigs["Name"]);
+        uint8_t fruId = static_cast<uint8_t>(fruConfigs["FruId"]);
+        std::shared_ptr<sdbusplus::asio::dbus_interface> iface =
+            objServer.add_interface(
+                "/xyz/openbmc_project/FruDevice/" + fruName,
+                "xyz.openbmc_project.Inventory.Item.FruConfig");
+
+        iface->register_property("Bus", busConf);
+        iface->register_property("Address", addrConf);
+        iface->register_property("FruId", fruId);
+        iface->initialize();
+    }
+}
 
 void rescanOneBus(
     BusMap& busmap, uint16_t busNum,
@@ -1457,6 +1500,11 @@ int main()
         rescanBusses(busMap, dbusInterfaceMap, unknownBusObjectCount, powerIsOn,
                      objServer, systemBus);
     });
+    /*TODO: Need to enable support once, D-Bus destruction is supported
+        iface->register_method("DoFruConfig", [&]() {
+            DoFruConfig(objServer);
+        });
+    */
     iface->initialize();
 
     std::function<void(sdbusplus::message_t & message)> eventHandler =
@@ -1554,7 +1602,7 @@ int main()
     // run the initial scan
     rescanBusses(busMap, dbusInterfaceMap, unknownBusObjectCount, powerIsOn,
                  objServer, systemBus);
-
+    DoFruConfig(objServer);
     io.run();
     return 0;
 }
-- 
2.34.1

From c577174b27ab0d51470904da4bd95971102b75ae Mon Sep 17 00:00:00 2001
From: dineshkumar <dineshkumar@ami.com>
Date: Thu, 1 Aug 2024 17:05:34 +0530
Subject: [PATCH] skip eeprom json check if file not found

---
 src/fru_device.cpp | 1 +
 1 file changed, 1 insertion(+)

diff --git a/src/fru_device.cpp b/src/fru_device.cpp
index 785e4b1..4541138 100644
--- a/src/fru_device.cpp
+++ b/src/fru_device.cpp
@@ -1023,6 +1023,7 @@ void DoFruConfig(sdbusplus::asio::object_server& objServer)
     if (!jsonFile.good())
     {
         std::cerr << "JSON file not found " << configFile << "\n";
+        return;
     }
 
     Json data = nullptr;
-- 
2.34.1

